devito/arguments.py

```diff
@@ -158,198 +107,350 @@ class PtrArgument(Argument):
-
+    def __repr__(self):
+        return "(" + self.dependency_type + ":" + str(self.obj) + ")"
 
-class ArrayArgProvider(ArgumentProvider):
 
-    """ Class used to decorate Tensor Function objects with behaviour required for runtime
-        arguments.
-    """
+class UnevaluatedDependency(object):
+    def __init__(self, consumer, evaluator, extra_param=None):
+        self.consumer = consumer
+        self.evaluator = evaluator
+        self.extra_param = extra_param
 
-    @cached_property
-    def rtargs(self):
-        return (TensorArgument(self.name, self),)
+    def evaluate(self, known_values):
+        return self.evaluator(self.consumer, known_values, self.extra_param)
 
 
-class ObjectArgProvider(ArgumentProvider):
-
-    """ Class used to decorate Objects with behaviour required for runtime arguments.
-    """
-
-    @cached_property
-    def rtargs(self):
-        return (PtrArgument(self.name, self),)
+def runtime_arguments(parameters):
+    return flatten([ArgumentVisitor().visit(p) for p in parameters])
 
 
 def log_args(arguments):
     arg_str = []
     for k, v in arguments.items():
         if hasattr(v, 'shape'):
-            arg_str.append('(%s, shape=%s, L2 Norm=%d)' %
-                           (k, str(v.shape), np.linalg.norm(v.view())))
+            arg_str.append('(%s, shape=%s, L2 Norm=%d, type=%s)' %
+                           (k, str(v.shape), np.linalg.norm(v.view()), type(v)))
         else:
-            arg_str.append('(%s, value=%s)' % (k, str(v)))
-    debug("Passing Arguments: " + ", ".join(arg_str))
-
-
-def infer_dimension_values_tuple(value, rtargs, offsets=None):
-    """ Strictly, a dimension's value is a 3-tuple consisting of the
-        values of each of its rtargs - currently size, start and end. However, for
-        convenience, we may accept partial representations of the value, e.g. scalars
-        and 2-tuples and interpret them in a certain way while assuming defaults for
-        missing information. If value is:
-        3-tuple: it contains explicit values for all 3 rtargs and hence will be used
-        directly
-        2-tuple: We assume we are being provided the (start, end) values. This will be
-        promoted to a 3-tuple assuming size to be the same as end.
-        scalar: We assume we are being provided the value of size. Promote to 3-tuple
-        by assuming this scalar is the size and the end of the dimension. start will
-        default to 0.
-    """
-    size_arg, start_arg, end_arg = rtargs
-    if not isinstance(value, tuple):
-        # scalar
-        value = (value, start_arg.default_value, value)
+            arg_str.append('(%s, value=%s, type=%s)' % (k, str(v), type(v)))
+    print("Passing Arguments: " + ", ".join(arg_str))
+
+
+def find_argument_by_name(name, haystack):
+    filtered = [(k, v) for k, v in haystack.items() if k.name == name]
+    assert(len(filtered) < 2)
+    if len(filtered) == 1:
+        return filtered[0][1]
```

Comment:
you could perhaps just store ``v`` in ``filtered``, so you won't need ``[1]`` here?